src/v0/destinations/mp/transform.js

```diff
@@ -470,28 +382,56 @@ const processRouterDest = async (inputs, reqMetadata) => {
     return errorRespEvents;
   }
 
-  const {
-    engageEventChunks,
-    groupsEventChunks,
-    trackEventChunks,
-    importEventChunks,
-    batchErrorRespList,
-  } = await processAndChunkEvents(inputs, reqMetadata);
-
-  const engageRespList = batchEvents(engageEventChunks, ENGAGE_MAX_BATCH_SIZE);
-  const groupsRespList = batchEvents(groupsEventChunks, GROUPS_MAX_BATCH_SIZE);
-  const trackRespList = batchEvents(trackEventChunks, TRACK_MAX_BATCH_SIZE);
-  const importRespList = batchEvents(importEventChunks, IMPORT_MAX_BATCH_SIZE);
-
-  let batchSuccessRespList = [
-    ...engageRespList,
-    ...groupsRespList,
-    ...trackRespList,
-    ...importRespList,
-  ];
-  batchSuccessRespList = combineBatchRequestsWithSameJobIds(batchSuccessRespList);
-
-  return [...batchSuccessRespList, ...batchErrorRespList];
+  const groupedEvents = groupEventsByType(inputs);
+  const response = await Promise.all(
+    groupedEvents.map(async (listOfEvents) => {
+      let transformedPayloads = await Promise.all(
+        listOfEvents.map(async (event) => {
+          try {
+            if (event.message.statusCode) {
+              // already transformed event
+              return {
+                message: event.message,
+                metadata: event.metadata,
+                destination: event.destination,
+              };
+            }
+
+            let processedEvents = await process(event);
+            processedEvents = CommonUtils.toArray(processedEvents);
+            return processedEvents.map((response) => ({
+              message: response,
+              metadata: event.metadata,
+              destination: event.destination,
+            }));
+          } catch (error) {
+            return handleRtTfSingleEventError(event, error, reqMetadata);
+          }
+        }),
+      );
+
+      transformedPayloads = _.flatMap(transformedPayloads);
+      const { engageEvents, groupsEvents, trackEvents, importEvents, batchErrorRespList } =
+        groupEventsByEndpoint(transformedPayloads);
+
+      const engageRespList = batchEvents(engageEvents, ENGAGE_MAX_BATCH_SIZE);
+      const groupsRespList = batchEvents(groupsEvents, GROUPS_MAX_BATCH_SIZE);
+      const trackRespList = batchEvents(trackEvents, TRACK_MAX_BATCH_SIZE);
+      const importRespList = batchEvents(importEvents, IMPORT_MAX_BATCH_SIZE);
+      const batchSuccessRespList = [
+        ...engageRespList,
+        ...groupsRespList,
+        ...trackRespList,
+        ...importRespList,
+      ];
+
+      return [...batchSuccessRespList, ...batchErrorRespList];
+    }),
+  );
+
+  // Flatten the response array containing batched events from multiple groups
+  const allBatchedEvents = _.flatMap(response);
+  return combineBatchRequestsWithSameJobIds(allBatchedEvents);
```

Comment:
Why should we do this? Why would jobIds repeat?

Reply:
In identify, group and track calls we are doing the multiplexing of events. So, two events can have the same JobIds.